python/paddle/distributed/auto_parallel/api.py

```diff
@@ -604,14 +598,56 @@ def __init__(self, optimizer, shard_fn=None):
             self._shard_clip = True
         self._inner_opt = optimizer
         self._shard_fn = shard_fn
+        self._sharding_mesh_axis = None
+        self._sharding_degree = None
 
-        # Invoke shard_fn if it is not None to shard parameters
-        if self._shard_fn is not None and isinstance(
-            self._shard_fn, ShardingStage3
-        ):
+        if isinstance(self._shard_fn, (ShardingStage1, ShardingStage3)):
+            self._set_and_check_sharding_prop_from_param()
+            self._shard_fn._set_sharding_mesh_axis(self._sharding_mesh_axis)
+
+        # Invoke shard_parameter in sharding stage 3 strategy
+        if isinstance(self._shard_fn, ShardingStage3):
             for param in self._inner_opt._parameter_list:
                 self._shard_fn._shard_parameter(param)
 
+    def _set_and_check_sharding_prop_from_param(self):
+        if len(self._shard_fn._mesh._shape) == 1:
+            self._sharding_degree = self._shard_fn._mesh.get_dim_size(0)
+            self._sharding_mesh_axis = 0
+        else:
+            param_list = self._inner_opt._parameter_list
+            for param in param_list:
+                if not param.is_dist():
+                    continue
+                mesh = param.process_mesh
+                placements = param.placements
+
+                if self._sharding_degree is None:
+                    # set the sharding degree if it has not been set
+                    if any(
+                        isinstance(placement, dist.Shard)
+                        for placement in placements
+                    ):
+                        for idx, placement in enumerate(placements):
+                            if isinstance(placement, dist.Replicate):
+                                self._sharding_degree = mesh.dim_size(idx)
+                                self._sharding_mesh_axis = idx
+                else:
+                    # check the sharding degree since it has already been set
+                    if any(
+                        isinstance(placement, dist.Shard)
+                        for placement in placements
+                    ):
+                        for idx, placement in enumerate(placements):
```

Comment:
check placement[self._sharding_mesh_axis] is replicated?

Reply:
done